backend/auto_operation/State.py

```diff
@@ -400,52 +481,42 @@ def getTrainInSection(self, section: Section) -> Train:
         else:
             return None
 
+    # 指定された地点に近づいてくる列車で一番近いものを取得
+    def getApproachingTrain(self, section: Section, mileage: float) -> Train:
+        testSection = section
+        testedSectionId = []  # 一度確認したセクションIDを記録しておき、2回通ったら抜けられないとみてNoneを返す
+        while True:
+            approachingTrain = self.getTrainInSection(testSection)
+            # testSectionに列車が存在しない、または存在しても引数で指定された地点より進んだ位置にいる（＝遠ざかっている）場合、探索セクションをひとつ前へ
+            if approachingTrain is None or (
+                approachingTrain.currentSection.id == section.id
+                and approachingTrain.mileage > mileage
+            ):
+                testedSectionId.append(testSection.id)  # 一度確認したセクションIDを記録しておく
+                testSection = testSection.sourceJunction.getInSection()  # ひとつ前のセクションに検索範囲を移す
+                if testSection.id in testedSectionId:  # すでに確認済みの場合、一周して戻ってしまったので、列車はいない。Noneを返す
+                    return None
+            # 存在すればその列車を返す
+            else:
+                return approachingTrain
+
     # 線路上のある点からある点までの距離を返す
-    # 2つの地点が同じsectionに存在する場合、s1>s2だと負の値を返す
-    def getDistance(
-        self,
-        s1: Section,
-        mileage1: float,
-        s2: Section,
-        mileage2: float,
-        originalStartSection: Section = None,
-    ) -> float:
+    # 2つの地点が同じsectionに存在する場合、s1>s2だと負の値を返す。ポイントの状態的に、ある点からある点にたどり着くのが不可能な場合、noneを返す
+    def getDistance(self, s1: Section, mileage1: float, s2: Section, mileage2: float) -> float:
         distance = 0
         testSection = s1
+        passedJunctionId: list[
+            int
+        ] = []  # 一度通過したjucntionのidを記録しておく。同じjunctionを2回通った場合は到達不能と判定しnanを返す
```

Comment:
```suggestion
       # 一度通過したjucntionのidを記録しておく。同じjunctionを2回通った場合は到達不能と判定しnanを返す
        passedJunctionId: list[int] = []
```
ここは多分こんな感じにしておくとフォーマッタを黙らせられます